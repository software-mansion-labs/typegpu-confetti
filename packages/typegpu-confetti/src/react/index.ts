export { default as Confetti } from './Confetti';

export {
  ConfettiProvider,
  useConfetti,
} from './ConfettiProvider';
