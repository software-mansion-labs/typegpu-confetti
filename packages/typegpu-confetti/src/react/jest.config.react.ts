import type { Config } from 'jest';

const config: Config = {
  displayName: 'react',
  testEnvironment: 'jest-environment-jsdom',
  rootDir: './../../',
  testMatch: ['<rootDir>/src/react/**/__tests__/**/*.test.{js,ts,tsx}'],
  setupFilesAfterEnv: ['<rootDir>/src/react/jest.setup.react.ts'],
  transform: {
    '^.+\\.tsx?$': [
      'ts-jest', 
      { 
        tsconfig: 'tsconfig.json', 
        babelConfig: true,
        diagnostics: false,
      }
    ],
  },
  moduleNameMapper: {
    '^src/(.*)$': '<rootDir>/src/$1',
  },
};

export default config;
