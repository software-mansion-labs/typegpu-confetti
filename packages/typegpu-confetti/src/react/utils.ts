import {
  useCallback,
  useContext,
  useEffect,
  useLayoutEffect,
  useReducer,
  useRef,
  useState,
} from 'react';
import type { AnyData, Infer } from 'typegpu/data';

import type { TgpuBuffer, TgpuRoot } from 'typegpu';
import { RootContext } from '../context';

export function useRoot(): TgpuRoot {
  const root = useContext(RootContext);

  if (root === null) {
    throw new Error('please provide root');
  }
  return root;
}

type State<T extends AnyData> = {
  buffer: TgpuBuffer<T>;
};
type Action<T extends AnyData> = { type: 'write'; value: Infer<T> };
type Initial<T extends AnyData> = {
  value: Infer<T> | undefined;
  label: string | undefined;
  root: TgpuRoot;
  schema: T;
};

function reducer<T extends AnyData>(
  state: State<T>,
  action: Action<T>,
): State<T> {
  if (action.type === 'write') {
    return state;
  }

  return state;
}

function createBuffer<T extends AnyData>({
  root,
  schema,
  label,
  value,
}: Initial<T>): State<T> {
  const buffer = root.createBuffer(schema, value).$name(label);

  return {
    buffer,
  };
}

export function useBuffer<T extends AnyData>(
  schema: T,
  value: Infer<T> | undefined,
  label?: string,
): TgpuBuffer<T> {
  const root = useRoot();

  const [{ buffer }] = useReducer(
    reducer,
    { root, schema, value, label },
    createBuffer<T>,
  );

  useEffect(() => {
    if (value !== undefined && buffer && !buffer.destroyed) {
      buffer.write(value);
    }
  }, [value, buffer]);

  const cleanupRef = useRef<ReturnType<typeof setTimeout> | null>(null);

  useEffect(() => {
    if (cleanupRef.current !== null) {
      clearTimeout(cleanupRef.current);
    }

    return () => {
      cleanupRef.current = setTimeout(() => {
        buffer.destroy();
      }, 1000);
    };
  }, [buffer]);

  return buffer;
}

// biome-ignore lint/suspicious/noExplicitAny: it's fine
function useEvent<TFunction extends (...params: any[]) => any>(
  handler: TFunction,
) {
  const handlerRef = useRef(handler);

  useLayoutEffect(() => {
    handlerRef.current = handler;
  });

  return useCallback((...args: Parameters<TFunction>) => {
    const fn = handlerRef.current;
    return fn(...args);
  }, []) as TFunction;
}

export function useFrame(
  loop: (deltaTime: number) => unknown,
  isRunning = true,
) {
  const loopEvent = useEvent(loop);
  useEffect(() => {
    if (!isRunning) {
      return;
    }

    let lastTime = Date.now();

    const runner = () => {
      const now = Date.now();
      const dt = now - lastTime;
      lastTime = now;
      loopEvent(dt);
      frame = requestAnimationFrame(runner);
    };

    let frame = requestAnimationFrame(runner);

    return () => cancelAnimationFrame(frame);
  }, [loopEvent, isRunning]);
}

interface DeviceContext {
  device: GPUDevice;
  adapter: GPUAdapter;
}

export const useDevice = (
  adapterOptions?: GPURequestAdapterOptions,
  deviceDescriptor?: GPUDeviceDescriptor,
) => {
  const [state, setState] = useState<DeviceContext | null>(null);

  useEffect(() => {
    setState(null); // resetting old adapter and device

    let deviceContext: Promise<DeviceContext> | DeviceContext = (async () => {
      const adapter = await navigator.gpu.requestAdapter(adapterOptions);
      if (!adapter) {
        throw new Error('No appropriate GPUAdapter found.');
      }
      const device = await adapter.requestDevice(deviceDescriptor);
      if (!device) {
        throw new Error('No appropriate GPUDevice found.');
      }
      deviceContext = { adapter, device };
      setState(deviceContext);
      return deviceContext;
    })();

    return () => {
      if (deviceContext instanceof Promise) {
        deviceContext.then((dev) => dev.device.destroy());
      } else {
        deviceContext.device.destroy();
      }
    };
  }, [adapterOptions, deviceDescriptor]);

  return { adapter: state?.adapter ?? null, device: state?.device ?? null };
};
