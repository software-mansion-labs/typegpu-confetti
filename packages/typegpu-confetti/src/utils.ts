import {
  useCallback,
  useContext,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import type { AnyData, Infer } from 'typegpu/data';

import type { TgpuBuffer, TgpuRoot } from 'typegpu';
import { RootContext } from './context';

export function useRoot(): TgpuRoot {
  const root = useContext(RootContext);

  if (root === null) {
    throw new Error('please provide root');
  }
  return root;
}

export function useBuffer<T extends AnyData>(
  schema: T,
  value: Infer<T> | undefined,
): TgpuBuffer<T> {
  const root = useRoot();
  const bufferRef = useRef<TgpuBuffer<T> | null>();

  // biome-ignore lint/correctness/useExhaustiveDependencies: <don't recreate buffer on value change>
  let buffer = useMemo(() => {
    if (bufferRef.current) {
      bufferRef.current.destroy();
    }
    const buffer = root.createBuffer(schema, value);
    bufferRef.current = buffer;
    return buffer;
  }, [root, schema]);

  // biome-ignore lint/style/noNonNullAssertion: <set in memo, strict mode acting weird>
  buffer = bufferRef.current!;

  useLayoutEffect(() => {
    if (value !== undefined && buffer && !buffer.destroyed) {
      buffer.write(value);
    }
  }, [value, buffer]);

  const cleanupRef = useRef<ReturnType<typeof setTimeout> | null>();

  useEffect(() => {
    if (cleanupRef.current !== null) {
      clearTimeout(cleanupRef.current);
    }

    return () => {
      cleanupRef.current = setTimeout(() => {
        buffer.destroy();
      }, 1000);
    };
  }, [buffer]);

  return buffer;
}

// biome-ignore lint/suspicious/noExplicitAny: it's fine
function useEvent<TFunction extends (...params: any[]) => any>(
  handler: TFunction,
) {
  const handlerRef = useRef(handler);

  useLayoutEffect(() => {
    handlerRef.current = handler;
  });

  return useCallback((...args: Parameters<TFunction>) => {
    const fn = handlerRef.current;
    return fn(...args);
  }, []) as TFunction;
}

export function useFrame(
  loop: (deltaTime: number) => unknown,
  isRunning = true,
) {
  const loopEvent = useEvent(loop);
  useEffect(() => {
    if (!isRunning) {
      return;
    }

    let lastTime = Date.now();

    const runner = () => {
      const now = Date.now();
      const dt = now - lastTime;
      lastTime = now;
      loopEvent(dt);
      frame = requestAnimationFrame(runner);
    };

    let frame = requestAnimationFrame(runner);

    return () => cancelAnimationFrame(frame);
  }, [loopEvent, isRunning]);
}

interface DeviceContext {
  device: GPUDevice;
  adapter: GPUAdapter;
}

export const useDevice = (
  adapterOptions?: GPURequestAdapterOptions,
  deviceDescriptor?: GPUDeviceDescriptor,
) => {
  const [state, setState] = useState<DeviceContext | null>(null);

  useEffect(() => {
    setState(null); // resetting old adapter and device

    let deviceContext: Promise<DeviceContext> | DeviceContext = (async () => {
      const adapter = await navigator.gpu.requestAdapter(adapterOptions);
      if (!adapter) {
        throw new Error('No appropriate GPUAdapter found.');
      }
      const device = await adapter.requestDevice(deviceDescriptor);
      if (!device) {
        throw new Error('No appropriate GPUDevice found.');
      }
      deviceContext = { adapter, device };
      setState(deviceContext);
      return deviceContext;
    })();

    return () => {
      if (deviceContext instanceof Promise) {
        deviceContext.then((dev) => dev.device.destroy());
      } else {
        deviceContext.device.destroy();
      }
    };
  }, [adapterOptions, deviceDescriptor]);

  return { adapter: state?.adapter ?? null, device: state?.device ?? null };
};
