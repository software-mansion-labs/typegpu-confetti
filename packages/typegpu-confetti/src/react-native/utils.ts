import { useState } from 'react';
import { PixelRatio } from 'react-native';
import { type RNCanvasContext, useCanvasEffect } from 'react-native-wgpu';
import { useRoot } from '../utils';

export function useGPUSetup(
  presentationFormat: GPUTextureFormat = navigator.gpu.getPreferredCanvasFormat(),
) {
  const root = useRoot();
  const [context, setContext] = useState<RNCanvasContext | null>(null);

  const ref = useCanvasEffect(() => {
    const ctx = ref.current?.getContext('webgpu');

    if (!ctx) {
      setContext(null);
      return;
    }

    const canvas = ctx.canvas as HTMLCanvasElement;
    canvas.width = canvas.clientWidth * PixelRatio.get();
    canvas.height = canvas.clientHeight * PixelRatio.get();

    ctx.configure({
      device: root.device,
      format: presentationFormat,
      alphaMode: 'premultiplied',
    });

    setContext(ctx);
  });

  return { canvasRef: ref, context };
}
