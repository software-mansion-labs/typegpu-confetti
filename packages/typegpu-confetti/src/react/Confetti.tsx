import React, {
  type CSSProperties,
  type ForwardedRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react';

import tgpu, {
  type TgpuComputePipeline,
  type TgpuRenderPipeline,
} from 'typegpu';
import * as d from 'typegpu/data';
import { RootContext } from '../context';
import { defaults } from '../defaults';
import {
  ParticleData,
  ParticleGeometry,
  addParticleCompute,
  canvasAspectRatio,
  dataLayout,
  deltaTime,
  geometryLayout,
  gravityFn,
  gravity as gravitySlot,
  initCompute,
  initParticleFn,
  initParticle as initParticleSlot,
  mainCompute,
  mainFrag,
  mainVert,
  maxDurationTime as maxDurationTimeSlot,
  maxParticleAmount as maxParticleAmountSlot,
  particles,
  time,
} from '../schemas';
import type { ConfettiPropTypes, ConfettiRef } from '../types';
import { useBuffer, useDevice, useFrame, useRoot } from '../utils';

const startTime = Date.now();

const ConfettiViz = React.forwardRef(
  (
    {
      width,
      height,
      gravity = defaults.gravity,
      colorPalette = defaults.colorPalette,
      initParticleAmount = defaults.initParticleAmount,
      maxParticleAmount: maxParticleAmount_ = defaults.maxParticleAmount,
      size = defaults.size,
      maxDurationTime = defaults.maxDurationTime,
      initParticle = defaults.initParticle,
      style = {},
    }: ConfettiPropTypes & { style?: CSSProperties } & {
      width: number;
      height: number;
    },
    ref: ForwardedRef<ConfettiRef>,
  ) => {
    const root = useRoot();

    const presentationFormat = navigator.gpu.getPreferredCanvasFormat();
    const canvasRef = useRef<HTMLCanvasElement | null>(null);
    const [context, setContext] = useState<GPUCanvasContext | null>(null);

    useEffect(() => {
      if (root && !context && canvasRef.current) {
        const ctx = canvasRef.current?.getContext('webgpu') as GPUCanvasContext;
        ctx.configure({
          device: root.device,
          format: presentationFormat,
          alphaMode: 'premultiplied',
        });
        setContext(ctx);
      }
    }, [context, root, presentationFormat]);

    const [ended, setEnded] = useState(false);
    const [timeoutKey, setTimeoutKey] = useState(0);

    const particleAmount = useRef(initParticleAmount);

    const maxParticleAmount = useMemo(
      () => Math.max(maxParticleAmount_, initParticleAmount, 0),
      [maxParticleAmount_, initParticleAmount],
    );

    // biome-ignore lint/correctness/useExhaustiveDependencies: <trigger timeout reset by changing timeoutKey>
    useEffect(() => {
      let timeout: ReturnType<typeof setTimeout> | undefined;
      if (maxDurationTime !== null) {
        timeout = setTimeout(
          () => setEnded(true),
          (maxDurationTime + 0.01) * 1000,
        );
      }
      return () => {
        if (timeout) {
          clearTimeout(timeout);
        }
      };
    }, [maxDurationTime, timeoutKey]);

    // #region buffers

    const canvasAspectRatioBuffer = useBuffer(
      d.f32,
      context ? context.canvas.width / context.canvas.height : 1,
    ).$usage('uniform');

    const canvasAspectRatioUniform = useMemo(
      () => canvasAspectRatioBuffer.as('uniform'),
      [canvasAspectRatioBuffer],
    );

    const particleGeometry = useMemo(
      () =>
        Array(maxParticleAmount)
          .fill(0)
          .map(() => ({
            angle: Math.floor(Math.random() * 50) - 10,
            tilt: (Math.floor(Math.random() * 10) - 20) * size,
            color: colorPalette.map(([r, g, b, a]) =>
              d.vec4f(r / 255, g / 255, b / 255, a),
            )[Math.floor(Math.random() * colorPalette.length)] as d.v4f,
          })),
      [colorPalette, maxParticleAmount, size],
    );

    const ParticleGeometryArray = useMemo(
      () => d.arrayOf(ParticleGeometry, maxParticleAmount),
      [maxParticleAmount],
    );
    const ParticleDataArray = useMemo(
      () => d.arrayOf(ParticleData, maxParticleAmount),
      [maxParticleAmount],
    );

    const particleGeometryBuffer = useBuffer(
      ParticleGeometryArray,
      particleGeometry,
    ).$usage('vertex');

    const particleDataBuffer = useBuffer(ParticleDataArray).$usage(
      'storage',
      'vertex',
    );

    const deltaTimeBuffer = useBuffer(d.f32).$usage('uniform');
    const timeBuffer = useBuffer(d.f32).$usage('storage');

    const particleDataStorage = useMemo(
      () => particleDataBuffer.as('mutable'),
      [particleDataBuffer],
    );
    const deltaTimeUniform = useMemo(
      () => deltaTimeBuffer.as('uniform'),
      [deltaTimeBuffer],
    );
    const timeStorage = useMemo(() => timeBuffer.as('readonly'), [timeBuffer]);

    //#endregion

    // #region pipelines

    const validatePipeline = useCallback(
      <T extends TgpuRenderPipeline | TgpuComputePipeline>(pipeline: T) => {
        root.device.pushErrorScope('validation');
        try {
          root.unwrap(pipeline as TgpuComputePipeline);
        } catch (error) {
          console.error(error);
          if (typeof error === 'object' && error && 'cause' in error) {
            console.log(error.cause);
          }
        }

        root.device.popErrorScope().then((error) => {
          if (error) {
            setEnded(true);
            console.error('error compiling pipeline', error.message);
          } else {
            // console.log('pipeline creation: no error');
          }
        });
        return pipeline;
      },
      [root],
    );

    const renderPipeline = useMemo(
      () =>
        validatePipeline(
          root['~unstable']
            .with(canvasAspectRatio, canvasAspectRatioUniform)
            .withVertex(mainVert, {
              tilt: geometryLayout.attrib.tilt,
              angle: geometryLayout.attrib.angle,
              color: geometryLayout.attrib.color,
              center: dataLayout.attrib.position,
              timeLeft: dataLayout.attrib.timeLeft,
            })
            .withFragment(mainFrag, {
              format: presentationFormat,
            })
            .withPrimitive({
              topology: 'triangle-strip',
            })
            .createPipeline(),
        ),
      [canvasAspectRatioUniform, presentationFormat, root, validatePipeline],
    );

    const computePipeline = useMemo(
      () =>
        validatePipeline(
          root['~unstable']
            .with(particles, particleDataStorage)
            .with(
              maxDurationTimeSlot,
              maxDurationTime ?? defaults.maxDurationTime,
            )
            .with(initParticleSlot, initParticleFn(initParticle))
            .with(gravitySlot, gravityFn(gravity))
            .with(time, timeStorage)
            .with(deltaTime, deltaTimeUniform)
            .withCompute(mainCompute)
            .createPipeline(),
        ),
      [
        particleDataStorage,
        root,
        timeStorage,
        gravity,
        validatePipeline,
        initParticle,
        maxDurationTime,
        deltaTimeUniform,
      ],
    );

    const initComputePipeline = useMemo(
      () =>
        validatePipeline(
          root['~unstable']
            .with(particles, particleDataStorage)
            .with(
              maxDurationTimeSlot,
              maxDurationTime ?? defaults.maxDurationTime,
            )
            .with(initParticleSlot, initParticleFn(initParticle))
            .with(time, timeStorage)
            .withCompute(initCompute)
            .createPipeline(),
        ),
      [
        particleDataStorage,
        root,
        maxDurationTime,
        validatePipeline,
        initParticle,
        timeStorage,
      ],
    );

    const addParticleComputePipeline = useMemo(
      () =>
        validatePipeline(
          root['~unstable']
            .with(particles, particleDataStorage)
            .with(
              maxDurationTimeSlot,
              maxDurationTime ?? defaults.maxDurationTime,
            )
            .with(initParticleSlot, initParticleFn(initParticle))
            .with(maxParticleAmountSlot, maxParticleAmount)
            .with(time, timeStorage)
            .withCompute(addParticleCompute)
            .createPipeline(),
        ),
      [
        particleDataStorage,
        root,
        maxParticleAmount,
        maxDurationTime,
        validatePipeline,
        initParticle,
        timeStorage,
      ],
    );

    // #endregion

    useImperativeHandle(
      ref,
      () =>
        ({
          pause: () => setEnded(true),
          resume: () => setEnded(false),
          restart: () => {
            particleAmount.current = initParticleAmount;

            if (initParticleAmount > 0) {
              initComputePipeline.dispatchWorkgroups(initParticleAmount);
            }

            if (ended) {
              setEnded(false);
            }
          },

          addParticles: (amount: number) => {
            for (let i = 0; i < amount; i++) {
              addParticleComputePipeline.dispatchWorkgroups(1);
            }

            particleAmount.current = Math.min(
              particleAmount.current + amount,
              maxParticleAmount,
            );

            if (ended) {
              setEnded(false);
            }
            setTimeoutKey((key) => key + 1);
          },
        }) satisfies ConfettiRef,
      [
        ended,
        maxParticleAmount,
        initParticleAmount,
        addParticleComputePipeline,
        initComputePipeline,
      ],
    );

    useEffect(() => {
      if (initParticleAmount > 0) {
        initComputePipeline.dispatchWorkgroups(initParticleAmount);
      }
    }, [initComputePipeline, initParticleAmount]);

    const frame = async (deltaTime: number) => {
      if (!context || particleAmount.current < 1) {
        return;
      }

      root.device.pushErrorScope('validation');

      deltaTimeBuffer.write(deltaTime);
      timeBuffer.write(Date.now() - startTime);
      canvasAspectRatioBuffer.write(
        context.canvas.width / context.canvas.height,
      );
      computePipeline.dispatchWorkgroups(
        Math.ceil(particleAmount.current / 64),
      );

      renderPipeline
        .with(geometryLayout, particleGeometryBuffer)
        .with(dataLayout, particleDataBuffer)
        .withColorAttachment({
          view: context.getCurrentTexture().createView(),
          clearValue: [0, 0, 0, 0],
          loadOp: 'clear',
          storeOp: 'store',
        })
        .draw(4, particleAmount.current);

      root['~unstable'].flush();

      root.device.popErrorScope().then((error) => {
        if (error) {
          console.error('error in loop', error.message);
          setEnded(true);
        }
      });
    };

    useFrame(frame, !ended);

    return (
      <canvas
        ref={canvasRef}
        width={width}
        height={height}
        style={{
          position: 'absolute',
          width: '100%',
          height: '100%',
          inset: 0,
          ...style,
        }}
      />
    );
  },
);

const Confetti = React.forwardRef(
  (props: ConfettiPropTypes, ref: ForwardedRef<ConfettiRef>) => {
    const { device } = useDevice();
    const root = useMemo(
      () => (device ? tgpu.initFromDevice({ device }) : null),
      [device],
    );

    const [height, setHeight] = useState(200);
    const [width, setWidth] = useState(200);

    const measuredRef = useRef<HTMLDivElement | null>(null);

    // biome-ignore lint/correctness/useExhaustiveDependencies: ref is null on first render
    useEffect(() => {
      if (measuredRef.current) {
        setWidth(
          measuredRef.current.getBoundingClientRect().width *
            window.devicePixelRatio,
        );
        setHeight(
          measuredRef.current.getBoundingClientRect().height *
            window.devicePixelRatio,
        );

        const observer = new ResizeObserver((e) => {
          setWidth((width) =>
            e[0] ? e[0].contentRect.width * window.devicePixelRatio : width,
          );
          setHeight((height) =>
            e[0] ? e[0].contentRect.height * window.devicePixelRatio : height,
          );
        });

        observer.observe(measuredRef.current);
      }
    }, [root]);

    if (root === null) {
      return null;
    }

    return (
      <RootContext.Provider value={root}>
        <div
          ref={measuredRef}
          style={{
            position: 'absolute',
            width: '100%',
            height: '100%',
            inset: 0,
            zIndex: 20,
            pointerEvents: 'none',
            cursor: 'auto',
          }}
        >
          <ConfettiViz {...props} ref={ref} width={width} height={height} />
        </div>
      </RootContext.Provider>
    );
  },
);

export default Confetti;
