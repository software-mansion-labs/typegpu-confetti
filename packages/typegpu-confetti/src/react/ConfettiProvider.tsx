import {
  type CSSProperties,
  type ReactNode,
  type RefObject,
  createContext,
  useContext,
  useRef,
} from 'react';
import type { ConfettiPropTypes, ConfettiRef } from '../types';
import Confetti from './Confetti';

const ConfettiContext = createContext<RefObject<ConfettiRef | null> | null>(
  null,
);

export function ConfettiProvider(
  props: { children: ReactNode, style?: CSSProperties } & ConfettiPropTypes,
) {
  const { children, ...confettiProps } = props;
  const ref = useRef<ConfettiRef>(null);

  return (
    <ConfettiContext.Provider value={ref}>
      <div style={{ width: '100%', height: '100%' }}>
        {children}

        <Confetti
          {...confettiProps}
          initParticleAmount={confettiProps.initParticleAmount ?? 0}
          maxParticleAmount={confettiProps.maxParticleAmount ?? 500}
          ref={ref}
        />
      </div>
    </ConfettiContext.Provider>
  );
}

export function useConfetti() {
  return useContext(ConfettiContext);
}
