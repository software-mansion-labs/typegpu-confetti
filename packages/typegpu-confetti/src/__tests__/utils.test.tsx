import type React from 'react';
import { renderHook, act } from '@testing-library/react';
import { RootContext } from '../context';
import { useBuffer } from '../utils';

// Mock the useRoot hook
const mockRoot = {
  createBuffer: jest.fn().mockReturnValue({
    write: jest.fn(),
    destroy: jest.fn(),
    destroyed: false,
  }),
};

// Create a wrapper component to provide the mocked context value
const wrapper = ({ children }: { children: React.ReactNode }) => (
  <RootContext.Provider value={mockRoot}>
    {children}
  </RootContext.Provider>
);

// Mock schema type for testing
const mockSchema = {
  type: 'buffer',
  size: 16,
} as any;

describe('useBuffer Hook', () => {
  beforeEach(() => {
    // Clear mock history before each test
    (mockRoot.createBuffer as jest.Mock).mockClear();
    (mockRoot.createBuffer().write as jest.Mock).mockClear();
    (mockRoot.createBuffer().destroy as jest.Mock).mockClear();
    jest.clearAllTimers();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it('should match snapshot for initial render', () => {
    const { result } = renderHook(() => useBuffer(mockSchema), { wrapper });
    
    expect({
      root: mockRoot,
      createBufferCalled: mockRoot.createBuffer.mock.calls.length,
      bufferMethods: {
        write: result.current.write,
        destroy: result.current.destroy,
        destroyed: result.current.destroyed,
      },
    }).toMatchSnapshot();
  });

  it('should match snapshot when value is provided', () => {
    const testValue = { x: 1, y: 2, z: 3 };
    const { result } = renderHook(() => useBuffer(mockSchema, testValue), {
      wrapper,
    });
    
    expect({
      root: mockRoot,
      createBufferCalled: mockRoot.createBuffer.mock.calls.length,
      providedValue: testValue,
      bufferMethods: {
        write: result.current.write,
        destroy: result.current.destroy,
        destroyed: result.current.destroyed,
      },
    }).toMatchSnapshot();
  });

  it('should match snapshot after value update', () => {
    const initialValue = { x: 1, y: 2 };
    const updatedValue = { x: 3, y: 4 };
    
    const { result, rerender } = renderHook(
      ({ value }) => useBuffer(mockSchema, value),
      {
        initialProps: { value: initialValue },
        wrapper,
      },
    );
    // Capture initial state
    const initialSnapshot = {
      createBufferCalls: mockRoot.createBuffer.mock.calls.length,
      writeCalls: (result.current.write as jest.Mock).mock?.calls?.length || 0,
    };

    // Update value
    act(() => {
      rerender({ value: updatedValue });
    });

    const finalSnapshot = {
      initial: initialSnapshot,
      afterUpdate: {
        createBufferCalls: mockRoot.createBuffer.mock.calls.length,
        writeCalls: (result.current.write as jest.Mock).mock?.calls?.length || 0,
        finalValue: updatedValue,
      },
    };

    expect(finalSnapshot).toMatchSnapshot();
  });

  it('should match snapshot for cleanup behavior', () => {
    const { result, unmount } = renderHook(() => useBuffer(mockSchema), {
      wrapper,
    });

    const beforeUnmount = {
      destroyCalled: (result.current.destroy as jest.Mock).mock?.calls?.length || 0,
      pendingTimers: jest.getTimerCount(),
    };

    act(() => {
      unmount();
    });

    const afterUnmount = {
      destroyCalled: (result.current.destroy as jest.Mock).mock?.calls?.length || 0,
      pendingTimers: jest.getTimerCount(),
    };

    // Fast-forward cleanup timeout
    act(() => {
      jest.advanceTimersByTime(1000);
    });

    const afterTimeout = {
      destroyCalled: (result.current.destroy as jest.Mock).mock?.calls?.length || 0,
      pendingTimers: jest.getTimerCount(),
    };

    expect({
      beforeUnmount,
      afterUnmount,
      afterTimeout,
    }).toMatchSnapshot();
  });

  it('should match snapshot when schema changes', () => {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    const schema1 = { type: 'buffer', size: 16 } as any;
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    const schema2 = { type: 'buffer', size: 32 } as any;
    
    const { result, rerender } = renderHook(
      ({ schema }) => useBuffer(schema),
      {
        initialProps: { schema: schema1 },
        wrapper,
      },
    );

    const initialState = {
      createBufferCalls: mockRoot.createBuffer.mock.calls.length,
      destroyCalls: (result.current.destroy as jest.Mock).mock?.calls?.length || 0,
    };

    act(() => {
      rerender({ schema: schema2 });
    });

    const finalState = {
      initial: initialState,
      afterSchemaChange: {
        createBufferCalls: mockRoot.createBuffer.mock.calls.length,
        destroyCalls: (result.current.destroy as jest.Mock).mock?.calls?.length || 0,
      },
    };

    expect(finalState).toMatchSnapshot();
  });
});