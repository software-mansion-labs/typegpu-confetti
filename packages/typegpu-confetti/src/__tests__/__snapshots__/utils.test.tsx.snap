// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`useBuffer Hook should match snapshot after value update 1`] = `
{
  "afterUpdate": {
    "createBufferCalls": 3,
    "finalValue": {
      "x": 3,
      "y": 4,
    },
    "writeCalls": 2,
  },
  "initial": {
    "createBufferCalls": 3,
    "writeCalls": 1,
  },
}
`;

exports[`useBuffer Hook should match snapshot for cleanup behavior 1`] = `
{
  "afterTimeout": {
    "destroyCalled": 1,
    "pendingTimers": 0,
  },
  "afterUnmount": {
    "destroyCalled": 0,
    "pendingTimers": 1,
  },
  "beforeUnmount": {
    "destroyCalled": 0,
    "pendingTimers": 0,
  },
}
`;

exports[`useBuffer Hook should match snapshot for initial render 1`] = `
{
  "bufferMethods": {
    "destroy": [MockFunction],
    "destroyed": false,
    "write": [MockFunction],
  },
  "createBufferCalled": 3,
  "root": {
    "createBuffer": [MockFunction] {
      "calls": [
        [],
        [],
        [
          {
            "size": 16,
            "type": "buffer",
          },
          undefined,
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": {
            "destroy": [MockFunction],
            "destroyed": false,
            "write": [MockFunction],
          },
        },
        {
          "type": "return",
          "value": {
            "destroy": [MockFunction],
            "destroyed": false,
            "write": [MockFunction],
          },
        },
        {
          "type": "return",
          "value": {
            "destroy": [MockFunction],
            "destroyed": false,
            "write": [MockFunction],
          },
        },
      ],
    },
  },
}
`;

exports[`useBuffer Hook should match snapshot when schema changes 1`] = `
{
  "afterSchemaChange": {
    "createBufferCalls": 4,
    "destroyCalls": 1,
  },
  "initial": {
    "createBufferCalls": 3,
    "destroyCalls": 0,
  },
}
`;

exports[`useBuffer Hook should match snapshot when value is provided 1`] = `
{
  "bufferMethods": {
    "destroy": [MockFunction],
    "destroyed": false,
    "write": [MockFunction] {
      "calls": [
        [
          {
            "x": 1,
            "y": 2,
            "z": 3,
          },
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  "createBufferCalled": 3,
  "providedValue": {
    "x": 1,
    "y": 2,
    "z": 3,
  },
  "root": {
    "createBuffer": [MockFunction] {
      "calls": [
        [],
        [],
        [
          {
            "size": 16,
            "type": "buffer",
          },
          {
            "x": 1,
            "y": 2,
            "z": 3,
          },
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": {
            "destroy": [MockFunction],
            "destroyed": false,
            "write": [MockFunction] {
              "calls": [
                [
                  {
                    "x": 1,
                    "y": 2,
                    "z": 3,
                  },
                ],
              ],
              "results": [
                {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
          },
        },
        {
          "type": "return",
          "value": {
            "destroy": [MockFunction],
            "destroyed": false,
            "write": [MockFunction] {
              "calls": [
                [
                  {
                    "x": 1,
                    "y": 2,
                    "z": 3,
                  },
                ],
              ],
              "results": [
                {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
          },
        },
        {
          "type": "return",
          "value": {
            "destroy": [MockFunction],
            "destroyed": false,
            "write": [MockFunction] {
              "calls": [
                [
                  {
                    "x": 1,
                    "y": 2,
                    "z": 3,
                  },
                ],
              ],
              "results": [
                {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;
